lib/Infer/AliveDriver.cpp

```diff
@@ -0,0 +1,411 @@
+  auto &bp = tgt_state.returnVal();
+  auto &a = ap.first;
+  auto &b = bp.first;
+
+  auto qvars = src_state.getQuantVars();
+  qvars.insert(ap.second.begin(), ap.second.end());
+
+  auto err = [&](const smt::Result &r, bool print_var, const char *msg) {
+    tools::error(errs, src_state, tgt_state, r, print_var, nullptr, a, b, msg,
+          false);
+  };
+
+  std::set<smt::expr> vars;
+  std::set<smt::expr> consts;
+
+  for (auto &[var, val] : src_state.getValues()) {
+    auto &name = var->getName();
+    if (StartsWith("%var", name)) {
+      auto app = val.first.value.isApp();
+      assert(app);
+      vars.insert(Z3_get_app_arg(smt::ctx(), app, 1));
+    }
+  }
+  for (auto &[var, val] : tgt_state.getValues()) {
+    auto &name = var->getName();
+    if (StartsWith("%reserved", name)) {
+      auto app = val.first.value.isApp();
+      assert(app);
+      consts.insert(Z3_get_app_arg(smt::ctx(), app, 1));
+    }
+  }
+
+  auto SimpleConstExistsCheck = smt::expr::mkForAll(vars, a.value == b.value);
+
+  smt::Solver::check({{preprocess(t, qvars, ap.second,
+                std::move(SimpleConstExistsCheck)),
+  [&] (const smt::Result &r) {
+    if (r.isUnsat()) {
+      err(r, true, "Value mismatch");
+    } else if (r.isSat()) {
+      result = r.getModel().getInt(*consts.begin());
+      return;
+    } else {
+      err(r, true, "Unknown/Invalid Result, investigate.");
+    }
+  }}});
+
+  return errs;
+}
+
+souper::AliveDriver::AliveDriver(Inst *LHS_, Inst *PreCondition_)
+    : LHS(LHS_), PreCondition(PreCondition_) {
+  InstNumbers = 101;
+
+  if (!TranslateRoot(LHS, PreCondition, LHSF, LExprCache)) {
+    llvm::errs() << "Failed to translate LHS.\n";
+    assert(false);
+  } else {
+    LHSFExists = true;
+  }
+}
+
+//TODO: Maybe return an APInt
+std::pair<int64_t, bool>
+souper::AliveDriver::synthesizeConstant(souper::Inst *RHS,
+                                        std::string ConstName) {
+  if (!LHSFExists) {
+    return {-1, false};
+  }
+
+  InstNumbers = 0;
+  std::set<std::string> Consts;
+  GetReservedConsts(RHS, Consts);
+  if (Consts.empty()) {
+    return {-1, verify(RHS)};
+  } else if (Consts.size() == 1) {
+    RExprCache.clear();
+    IR::Function RHSF;
+    if (!TranslateRoot(RHS, nullptr, RHSF, RExprCache)) {
+      llvm::errs() << "Failed to translate RHS.\n";
+      // TODO: Eventually turn this into an assertion
+      return {-1, false};
+    }
+    tools::Transform t;
+    ReturnLHSRAII foo{t, LHSF};
+    t.src = std::move(LHSF);
+    t.tgt = std::move(RHSF);
+    tools::TransformVerify tv(t, /*check_each_var=*/false);
+    int64_t result = 0;
+    if (auto errs = SynthesizeConstantUsingSolver(t, ConstName, result)) {
+      return {-1, false}; // TODO: Encode errs into ErrorCode
+    } else {
+      return {result, true};
+    }
+  } else {
+    assert(false && "Multiple constants unimplemented.");
+    return {-1, false};
+  }
+}
+
+
+bool souper::AliveDriver::verify (Inst *RHS) {
+  if (!LHSFExists) {
```

Comment:
we do this both ways, so it you're free to choose either one, but in some cases like this it may make the code more readable to leave out the curly braces